pilot/pkg/xds/ads.go

```diff
@@ -161,23 +161,26 @@ func isExpectedGRPCError(err error) bool {
 func receiveThread(con *Connection, reqChannel chan *discovery.DiscoveryRequest, errP *error) {
 	defer close(reqChannel) // indicates close of the remote side.
 	for {
+		con.mu.RLock()
+		cid := con.ConID
```

Comment:
nit: move this outside the loop

Reply:
Won't work - that's the point of the mutex, con.ConID is set in a different go routine - the one handling the reqChannel and calling initConnection. On the first call guaranteed to not be set - I moved it bellow Recv() - seems better.

The second packet received - if the timing is right, and the first packet is processed - will have it set. Only used for logging, so not a huge deal. Maybe we should just not log it and clean up the mutex..

Reply:
Could we move the connection init into this function

Reply:
Good idea - it's really the first packet in the stream that matters, and that may simplify the code.

However: I already had to split a large PR in 2 parts, how about we move this in a separate PR, before this PR becomes too large :-) It is certainly a good idea to do the init in the receive thread. I'm thinking even more.

Reply:
https://github.com/istio/istio/pull/25124

Reply:
Ahh got it. forgot about that. The other PR lgtm